pkg/kubelet/kubelet_pods.go

```diff
@@ -2328,3 +2361,68 @@ func (kl *Kubelet) cleanupOrphanedPodCgroups(pcm cm.PodContainerManager, cgroupP
 		go pcm.Destroy(val)
 	}
 }
+
+func (kl *Kubelet) runtimeClassSupportsRecursiveReadOnlyMounts(pod *v1.Pod) bool {
+	var runtimeClassName string
+	if pod.Spec.RuntimeClassName != nil {
+		runtimeClassName = *pod.Spec.RuntimeClassName
+	}
+	rtClasses := kl.runtimeState.runtimeClasses()
+	return runtimeClassSupportsRecursiveReadOnlyMounts(runtimeClassName, rtClasses)
+}
+
+// runtimeClassSupportsRecursiveReadOnlyMounts checks whether the runtime class supports recursive read-only mounts.
+// The kubelet feature gate is not checked here.
+func runtimeClassSupportsRecursiveReadOnlyMounts(runtimeClassName string, runtimeClasses []v1.RuntimeClass) bool {
+	for _, f := range runtimeClasses {
+		if f.Name == runtimeClassName {
+			if f.Features == nil {
+				return false
+			}
+			if f.Features.RecursiveReadOnlyMounts == nil {
+				return false
+			}
+			if !*f.Features.RecursiveReadOnlyMounts {
+				return false
+			}
+			return true
+		}
+	}
+	klog.ErrorS(nil, "unknown runtime class", "runtimeClassName", runtimeClassName)
+	return false
+}
+
+// resolveRecursiveReadOnly resolves the recursive read-only mount mode.
+func resolveRecursiveReadOnly(m v1.VolumeMount, runtimeSupportsRRO, featureGateEnabled bool) (bool, error) {
+	if m.RecursiveReadOnly == nil {
+		return false, nil
+	}
+	switch rroMode := *m.RecursiveReadOnly; rroMode {
+	case v1.RecursiveReadOnlyDisabled:
+		return false, nil
+	case v1.RecursiveReadOnlyEnabled:
+		if !featureGateEnabled {
+			return false, fmt.Errorf("volume %q requested recursive read-only mode, but it is not enabled by kubelet feature gate (%q)",
+				m.Name, features.RecursiveReadOnlyMounts)
+		}
+		if !runtimeSupportsRRO {
+			return false, fmt.Errorf("volume %q requested recursive read-only mode, but it is not supported by the runtime", m.Name)
+		}
+		if !m.ReadOnly {
+			return false, fmt.Errorf("volume %q requested recursive read-only mode, but it is not read-only", m.Name)
+		}
+		if m.MountPropagation != nil && *m.MountPropagation != v1.MountPropagationNone {
+			return false, fmt.Errorf("volume %q requested recursive read-only mode, but it is not compatible with propagation %q",
+				m.Name, *m.MountPropagation)
+		}
+		return true, nil
+	case v1.RecursiveReadOnlyIfPossible:
+		propNone := true
+		if m.MountPropagation != nil {
+			propNone = *m.MountPropagation == v1.MountPropagationNone
+		}
+		return featureGateEnabled && runtimeSupportsRRO && m.ReadOnly && propNone, nil
```

Comment:
very picky nit: for some reason my brain wants this to be between disabled and enabled. I don't see it effecting the code flow. Feel free to ignore, it is of no consequence

Reply:
fixed